---
title: "DADA2 + PacBio: ZymoBIOMICS Microbial Community Standard"
author: "Benjamin J Callahan"
output: html_document
---

## Data

The mock community being sequenced is the ZymoBIOMICS Microbial Community Standard: https://www.zymoresearch.com/zymobiomics-community-standard . This mock community contains the following bacterial and yeast species (but we'll ignore the yeast as they aren't amplified by this protocol):

* Pseudomonas aeruginosa
* Escherichia coli
* Salmonella enterica
* Lactobacillus fermentum
* Enterococcus faecalis
* Staphylococcus aureus
* Listeria monocytogenes
* Bacillus subtilis

Of note, Zymo Research replaced five strains in the ZymoBIOMICS™ standards with similar strains beginning with Lot ZRC190633. The Lot # of the sample analyzed here was ZRC187325, which contains the “old” mixture of strains. For indiscernible reasons that suggest caution in ordering microbial "standards" products from this company, Zymo Research will not share the identify of all the strains in the older products. However, they did confirm the identify of the E. coli strain in the old products as E. coli O157:H7 str. CDC B6914-MS1.

Saving the 16S copy numbers for the bacteria from the table:
```{r}
ncopy <- c("Pseudomonas"=4, "Escherichia"=7, "Salmonella"=7, "Lactobacillus"=5, 
             "Enterococcus"=4, "Staphylococcus"=6, "Listeria"=6, "Bacillus"=10)
```

This sequencing data was generated by PacBio using their in-house 16S amplification and sequencing protocol: https://www.pacb.com/wp-content/uploads/Unsupported-Full-Length-16S-Amplification-SMRTbell-LibraryPreparation-and-Sequencing.pdf

The sequencing was performed on a Sequel, with the S/P1-C1.2 sequencing chemistry. Standard PacBio bioinformatics were performed, and consensus sequences were extracting using CCS 3.1.1 with default parameters, `minPasses=3` and `minPredictedAccuracy=0.999`:

> `ccs --pbi --force --logLevel=DEBUG --numThreads=16 --minSnr=3.75 --minReadScore=0.65 --maxLength=7000 --minLength=10 --minPasses=3 --minZScore=-5 --maxDropFraction=0.34 --minPredictedAccuracy=0.999 subreads.bam ccs.bam`

## Setup

Load libraries, setup paths, prepare environment:
```{r init, warning=FALSE, message=FALSE}
library(dada2);packageVersion("dada2")
library(Biostrings);packageVersion("Biostrings")
library(ShortRead);packageVersion("ShortRead")
library(ggplot2);packageVersion("ggplot2")
library(reshape2);packageVersion("reshape2")
library(RColorBrewer);packageVersion("RColorBrewer")
path <- "~/Desktop/LRAS/Data/Zymo/" # CHANGE ME to location of the fastq file
path.out <- "Figures/" 
path.rds <- "RDS/" 
fn <- file.path(path, "zymo_CCS_99_9.fastq.gz")
F27 <- "AGRGTTYGATYMTGGCTCAG"
R1492 <- "RGYTACCTTGTTACGACTT"
rc <- dada2:::rc
theme_set(theme_bw())
genusPalette <- c(Bacillus="#e41a1c", Enterococcus="#377eb8", Escherichia="#4daf4a", Lactobacillus="#984ea3",
                  Listeria="#ff7f00", Pseudomonas="#ffff33", Salmonella="#a65628", Staphylococcus="#f781bf")
```

## Remove Primers and Filter

Remove primers and orient reads:
```{r primers}
nop <- file.path(path, "noprimers", basename(fn))
prim <- removePrimers(fn, nop, primer.fwd=F27, primer.rev=dada2:::rc(R1492), orient=TRUE, verbose=TRUE)
```

Very little loss there even though primer indels aren't allowed currently. 

Inspect length distribution.
```{r}
hist(nchar(getSequences(nop)), 100)
```

Sharply peaked at the expected length range.

Filter:
```{r filter}
filt <- file.path(path, "noprimers", "filtered", basename(fn))
track <- fastqFilter(nop, filt, minQ=3, minLen=1000, maxLen=1600, maxN=0, rm.phix=FALSE, maxEE=2, verbose=TRUE)
```

Very little lost to filtering.

## Run DADA2

Dereplicate:
```{r derep}
drp <- derepFastq(filt, verbose=TRUE)
```

Learn errors:

```{r learn-err}
err <- learnErrors(drp, BAND_SIZE=32, multithread=TRUE, errorEstimationFunction=dada2:::PacBioErrfun) # 10s of seconds
```

Inspect errors:
```{r plot-err}
plotErrors(err)
```

Looks good.

Denoise:
```{r dada}
dd <- dada(drp, err=err, BAND_SIZE=32, multithread=TRUE) # seconds
dd
```

Read tracking:
```{r}
cbind(ccs=prim[,1], primers=prim[,2], filtered=track[[2]], denoised=sum(dd$denoised))
```

Inspect clustering data.frame
```{r inspect}
dd$clustering[,-1]
```

Very strong differentiation of true sequences and errors based on the classification p-values reported.

## Taxonomy and Chimeras

Assign taxonomy:
```{r taxonomy}
tax <- assignTaxonomy(dd, "~/tax/silva_nr_v128_train_set.fa.gz", multithread=TRUE)
tax[,"Genus"] <- gsub("Escherichia/Shigella", "Escherichia", tax[,"Genus"]) # Reformat to be compatible with other data sources
unname(tax)
```

No yeasts, presumably lost due to the bacterial primers, but all the expected bacterial genera are there.

Check chimeras:
```{r chimeras}
bim <- isBimeraDenovo(dd, minFoldParentOverAbundance=3.5, multithread=TRUE) 
# Higher MFPOA to avoid flagging intra-genomic variants
table(bim)
```

No chimeras!

Save processed objects for future analysis.
```{r saveRDS}
saveRDS(dd, file.path(path.rds, "Zymo_dd.rds"))
saveRDS(tax, file.path(path.rds, "Zymo_tax_Silva128.rds"))
```

Reload processed data objects (can run code below from here in absence of input sequence data):
```{r readRDS}
dd <- readRDS(file.path(path.rds, "Zymo_dd.rds"))
tax <- readRDS(file.path(path.rds, "Zymo_tax_Silva128.rds"))
```

## Reference check againt nt

BLAST these 29 sequences against nt (https://blast.ncbi.nlm.nih.gov/Blast.cgi):
```{r}
## dada2:::pfasta(dd, id=tax[,6])
uniquesToFasta(dd, file.path(path.rds, "Zymo_ASVs.fa"), ids=paste0(tax[,"Genus"], ";size=", dd$denoised, ";"))
is.ec <- tax[,"Genus"] %in% "Escherichia"
uniquesToFasta(dd$clustering[is.ec,], file.path(path.rds, "Zymo_Ecoli.fa"), ids=paste0("Ecoli", seq(sum(is.ec)), "_Zymo"))
## BLAST against nt excluding uncultured/environmental sequences (checkbox)
```

Results of BLAST search on July 18, 2018 are recorded.

 1. Exact match (100% identity, 100% coverage) to S. enterica strains: 2014LSAL02547, 08-00436, SA20035215, C500, CFSAN002050, NBRC 105684, JCM 1651
 2. Exact match to  (~10) B. subtilis/xiamenensis/instestinalis strains, including NRS 231 and ATCC 6633
 3. Exact match to many (>100) S. aureus strains.
 4. Exact match to many (>100) L. monocytogenes strains.
 5. Exact match to many (~40) L. monocytogenes strains.
 6. Exact match to many (~50) E. faecalis strains.
 7. Exact match to B. subtilis/intestinalis strains: T30, NRS 231, and W23
 8. Exact match to many (>100) Pseudomonas aeruginosa strains.
 9. Exact match to many (~50) E. coli strains.
10. Exact match to E. faecalis strains: KB1, D32, SE-8, HN-S5
11. Exact match to E. faecalis strain: NRIC 0110
12. Exact match to S. aureus strains: NCTC9944, MRSA107, TOHH628, V605, V521, Gv88, TW20
13. *One mismatch (1st base) to many S. aureus strains.*
14. Exact match to many S. aureus strains (99).
15. *One mismatch to L. fermentum strains: 3872, FTDC 8312, HBUAS54017*
16. Exact match to L. fermentum strains: CBA7106, FTDC 8312, 3872
17. Exact match to many (~20) L. fermentum strains.
18. Exact match to many (~15) L. monocytogenes strains.
19. Exact match to B. subtilis/intestinalis strains: T30, NRS 231, W23
20. *One mismatch to many (~25) L. fermentum strains.*
21. Exact match to B. subtilis/intestinalis strains: T30, NRS 231, W23
22. *One mismatch to ~10 B. subtilis/xiamenensis/instestinalis strains, including NRS 231 and ATCC 6633.*
23. *One mismatch AND one 10nt insertion to Lactobacillus fermentum 3872*
24. Exact match to B. subtilis strains: NRS 231, W23
25. *Two mismatches to many (~20) S. enterica strains.*
26. Exact match to (~10) E. coli strains.
27. Exact match to many (>100) E. coli strains.
28. Exact match to many (~20) E. coli strains.
29. Exact match to (~15) E. coli strains.

Everything is an exact match except for: 13, 15, 20, 22, 23, 25.
13 (Staphylococcus), 15/20/23 (Lactobacillus), 22 (Bacillus) and 25 (Salmonella).

Also 11 (Enterococcus) matches just 1 incomplete strain in NCBI nr/nt. The rest all match multiple.

Store these results in R format:
```{r}
refhit.nt <- c(TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE,
               TRUE, TRUE, FALSE,TRUE, FALSE,TRUE, TRUE, TRUE, TRUE, FALSE,
               TRUE, FALSE,FALSE,TRUE, FALSE,TRUE, TRUE, TRUE, TRUE)
which(!refhit.nt)
```

Correctly recorded.

## Reference check against provided 16S references

Zymo provides 16S reference sequences for these strains. They can be downloaded from the manufacturer's website. However, these references are sadly lacking. Remember, this is "Version 1" of the Microbial standard. However, the version 1 16S reference sequences are totally incomplete, typically containing just a single 16S sequence per organism. The "Version 2" references contain a full complement of 16S sequences for each strain, but 5/8 strains are wrong, including E. coli! Our best option is to use Version 2, and then combine with the nt database hits, and simply be honest that what we are describing are (mostly exact) matches to previously observed sequences from these species, rather than matches to an authoritative references which does not exist.

Import the 16S references and compare the recovered ASVs:
```{r}
path.16S <- file.path(path, "RefSeqs/ZymoBIOMICS.STD.refseq.v2/ssrRNAs") ### Version 2, 5/8 strains are wrong
### path.16S <- file.path(path, "RefSeqs/BioPool_genomes/16S-18S") #### Version 1, just 1 allele per strain
fn16S <- list.files(path.16S, pattern=".fasta$")
names(fn16S) <- sapply(strsplit(fn16S, "_"), `[`, 1)
fn16S <- fn16S[names(ncopy)] # Drop yeast
ncopy.ref <- sapply(file.path(path.16S, fn16S), function(x) length(getSequences(x)))
names(ncopy.ref) <- names(fn16S)
ncopy.ref
```

All 8 strains are present in the reference files, and have the expected number of 16S sequences.

Compare the sequences to all the reference sequences, allowing no mismatches (but with `fixed=FALSE` so as to match ambiguous nucleotides in the references):
```{r}
sq <- dd$sequence
ref.16S <- DNAStringSet(unlist(sapply(file.path(path.16S, fn16S), getSequences)))
names(ref.16S) <- sapply(strsplit(names(ref.16S), "[.]fasta[.]"), `[`, 2)
names(ref.16S) <- gsub("\\sConcatenation of 2 sequences", "", names(ref.16S))
refmat.16S <- sapply(DNAStringSet(sq), vcountPattern, subject=ref.16S, fixed=TRUE)
rownames(refmat.16S) <- names(ref.16S)
colnames(refmat.16S) <- paste0("Seq", seq_along(sq))
```

Inspect:
```{r}
refhit.16S <- colSums(refmat.16S)>0
which(refhit.16S)
table(tax[,"Genus"], Reference.Match=refhit.16S)
```

They mostly match the 16S references provided, although some don't at all (e.g. E. coli, which we know is a different strain in our version of the standard than in these references sequences). *Note though that nt non-hits Seq22 and Seq25 match here.*

## Consolidated reference comparisons

We'll now consider the total comparisons to all three references sources: nr/nt, Zymo-provided 16S, and Zymo-provided genomes.

```{r}
refmat <- matrix(c(refhit.nt, refhit.16S), ncol=2)
colnames(refmat) <- c("nr.nt", "zymo.16S")
rownames(refmat) <- paste0("Seq", seq_along(sq))
refmat
```

Consider whether the sequences appear, exaclty, in any of the reference sources:
```{r}
refhit.any <- rowSums(refmat) > 0
refhit.any
```

Which don't have any hits:
```{r}
refhit.any[!refhit.any]
unname(tax[!refhit.any, "Genus"])
```

All the rest have exact reference hits. The non-hits are mostly the 3/5 *L. fermentum* sequences, and also one *S. aureus* sequence variant.

## Intra-genomic Stoichiometry

Now we'll investigate the "stoichiometry" of the sequence variants from each bacterial strain, i.e. the copy number of each 16S sequence variants relative to the genomic copy number, all inferred from the dada2 processed data and the rRNA gene copy numbers given for each strain in the Zymo reference.

First calculate the genomic abundances, by summing over all sequence variants for each genus and dividing by the 16S copy number:
```{r}
abund.ome <- sapply(names(ncopy), function(gen) {
  is.gen <- grepl(gen, tax[,"Genus"])
  sum(dd$denoised[is.gen])/ncopy[gen]
})
names(abund.ome) <- names(ncopy)
dfgen <- data.frame(Genus=names(ncopy), Abundance=abund.ome, stringsAsFactors = FALSE)
ggplot(data=dfgen, aes(x=Genus, y=Abundance)) + 
  geom_col(width=0.4, aes(fill=Genus)) + scale_fill_manual(values=genusPalette) +
  ylim(c(0, NA)) + geom_hline(yintercept=mean(abund.ome), linetype="dashed") +
  theme(axis.text.x=element_blank(), axis.ticks.x = element_blank()) +
  ylab("Genome Abundance")
```

They are supposed to be all equal. Clearly they aren't, but all are present at reasonable frequencies (i.e. <3x abundance biases).

Now we make a similar plot for the abundance of each sequence variant:
```{r}
dfasv <- data.frame(Genus=tax[,"Genus"], Abundance=dd$denoised, stringsAsFactors = FALSE)
rownames(dfasv) <- NULL
ggplot(data=dfasv, aes(x=Genus, y=Abundance)) + 
  geom_point(aes(color=Genus), shape="x", size=4) + scale_color_manual(values=genusPalette) +
  ylim(c(0, NA)) +
  theme(axis.text.x=element_blank(), axis.ticks.x = element_blank()) +
  ylab("ASV Abundance")
```

And now we make the "stoichiometry" figure, i.e the abundance of each ASV scaled to the genomic abundance:
```{r}
dfasv$ScaledAbundance <- dfasv$Abundance/abund.ome[dfasv$Genus]
# Number the ASVs in each strain/genus
dfasv$Variant <- sapply(seq(nrow(dfasv)), function(i) sum(dfasv$Genus[1:i] == dfasv$Genus[[i]], na.rm=TRUE))

p.stoich <- ggplot(data=dfasv, aes(x=Variant, y=ScaledAbundance, fill=Genus, width=0.5)) + geom_col() + 
  scale_fill_manual(values=genusPalette) +
  facet_wrap(~Genus, nrow=2) +
  scale_y_continuous(breaks=seq(0,round(max(dfasv$ScaledAbundance))), minor_breaks=NULL) +
  theme(panel.grid.major.y=element_line(color="grey60", size=0.2)) +
  theme(panel.grid.major.x=element_blank(), panel.grid.minor.x=element_blank()) +
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()) +
  xlab("Full-length 16S Sequence Variants") + 
  ylab("Abundance (per-genome)") + 
  guides(fill=FALSE)
p.stoich
ggsave(file.path(path.out, "Zymo_Stoich.pdf"), p.stoich, width=178, height=100, units="mm", useDingbats=FALSE)
```

That is beautiful. There is an integer number of each sequence variant in each genome, just as there must be. That is, these aren't sequencing errors, these are real sequence variants. All of them.

Make another figure that compares the integral values to those if we change genome copy number to be some random-ish value:
```{r}
set.seed(100)

dfasv$Ncopy <- dfasv$ScaledAbundance
# Function. Genus. Named vector with genome copy numbers. Assumes dd and tax in outer environment.
get.abund.ome <- function(genera, ncopies) {
  rval <- sapply(genera, function(gen) {
    is.gen <- grepl(gen, tax[,"Genus"])
    sum(dd$denoised[is.gen])/ncopies[gen]
  })
  names(rval) <- genera
  rval
}
# Permute abund.ome
foo <- sample(length(ncopy))
while(any(foo == seq(length(ncopy)))) foo <- sample(length(ncopy)) # Ensure all switch
abund.perm <- abund.ome; abund.perm[1:length(abund.perm)] <- abund.perm[foo]
dfasv$PermuteAbundance <- dfasv$Abundance/abund.perm[dfasv$Genus]
# Permute copy numbers
foo <- sample(length(ncopy))
while(any(foo == seq(length(ncopy)))) foo <- sample(length(ncopy)) # Ensure all switch
ncopy.perm <- ncopy; ncopy.perm[1:length(ncopy.perm)] <- ncopy.perm[foo]
abund.ncopy.perm <- get.abund.ome(names(ncopy), ncopy.perm)
dfasv$PermuteNcopy <- dfasv$Abundance/abund.ncopy.perm[dfasv$Genus]
# Set copy numbers to the number of unique sequence variants
nasv <- table(tax[,"Genus"])
abund.nasv <- get.abund.ome(names(ncopy), nasv)
dfasv$NasvsNcopy <- dfasv$Abundance/abund.nasv[dfasv$Genus]
# Set copy numbers to the median copy number
ncopy.median <- ncopy; ncopy.median[1:length(ncopy.median)] <- median(ncopy)
abund.ncopy.median <- get.abund.ome(names(ncopy), ncopy.median)
dfasv$MedianNcopy <- dfasv$Abundance/abund.ncopy.median[dfasv$Genus]

df4 <- melt(dfasv, measure.vars=c("Ncopy", "PermuteAbundance", "PermuteNcopy", "NasvsNcopy", "MedianNcopy"), 
            value.name="Count", variable.name="Method")

ggplot(data=df4, aes(x=Method, y=Count, color=Genus)) + geom_jitter(height=0, width=0.15) +
  theme_bw() + theme(panel.grid.major.y=element_line(color="black", size=0.25)) +
#  theme(aspect.ratio=10) + 
#  scale_x_continuous(limits=c(0.7,1.3), breaks=NULL) +
  scale_y_continuous(breaks=seq(round(max(df4$Count))), minor_breaks=NULL)
  
```

It is only with the true copy numbers that we get the integer ladder.

## PacBio Quality Diagnostics

We now consider the distribution of PacBio CCS sequencing errors, using the reconstructed ASVs as the ground truth as confirmed above.

Rerun dada correcting all reads:
```{r}
dda <- dada(drp, err=err, BAND_SIZE=32, OMEGA_C=0, multi=TRUE)
dda
```

Record all errors between the raw reads and the true ASVs by position (1-1500ish) and by type (D:Deletion, I:Insertion, S:Substitution).
```{r}
library(Biostrings)
diffs <- function(query, ref, vec=TRUE, ...) { # Currently not capturing insertions
  al <- nwalign(query, ref, vec=vec, ...)
  cstr <- compareStrings(al[[1]], al[[2]])
  cstr.ref <- gsub("[+]", "", cstr) # Ref coordinates, but lose insertion info
  cstr.q <- gsub("[-]", "", cstr) # Query coordinates, but lose deletion info
  cstr.ins <- gsub("[+]+", "+", cstr) # Reduce all multi-base inserts to a single insert
  cstr.del <- gsub("[-]+", "-", cstr) # Reduce all multi-base deletions to a single deletion
  refpos.sub <- unlist(gregexpr(pattern='[?]', cstr.ref))
  refpos.del <- unlist(gregexpr(pattern='[-]+', cstr.ref)) # Just getting loc of first deleted base of multi-nt deletions
  refpos.ins <- unlist(gregexpr(pattern='[+]', cstr.ins))
  refpos.ins <- refpos.ins - seq_along(refpos.ins) + 1 # Correct back to ref coords
  qpos.sub <- unlist(gregexpr(pattern='[?]', cstr.q))
  qpos.ins <- unlist(gregexpr(pattern='[+]+', cstr.q)) # Just getting loc of first inserted base of multi-nt inserts
  qpos.del <- unlist(gregexpr(pattern='[-]', cstr.del))
  qpos.del <- qpos.del - seq_along(qpos.del) + 1 # Correct back to ref coords
  rv <- rbind( data.frame(Type="S", RefPos=refpos.sub, QueryPos=qpos.sub),
               data.frame(Type="D", RefPos=refpos.del, QueryPos=qpos.del),
               data.frame(Type="I", RefPos=refpos.ins, QueryPos=qpos.ins))
  rv[rv$RefPos > -1,]
}

df.diffs <- function(i) {
  df <- diffs(getSequences(drp)[i], dd$sequence[dd$map[i]])
  df <- cbind(df, Abund = rep(drp$uniques[i], nrow(df)), 
              Derep=rep(i, nrow(df)), Denoised=rep(dd$map[i], nrow(df)))
  df$Qual <- drp$quals[cbind(df$Derep, df$QueryPos)]
  df
}
```

Identify chimeras in the denoised reads:
```{r}
bim.drp <- isBimeraDenovo(drp, minFoldParentOverAbundance=3.5, multi=TRUE) # A couple minutes
table(bim.drp)
unname(drp$uniques[unname(head(which(bim.drp)))])
```

Identify which uniques were corrected/uncorrected:
```{r}
corrected.drp <- !is.na(dd$map)
table(corrected.drp)
```

Identify contaminants as those reads that are not assigned to one of the genera in the mock community.
```{r}
tax.drp.uncorrected <- assignTaxonomy(getSequences(drp)[!corrected.drp], "~/tax/silva_nr_v128_train_set.fa.gz", multi=TRUE) # 
table(tax.drp.uncorrected[,"Genus"], useNA="ifany")
table(bim.drp[!corrected.drp][is.na(tax.drp.uncorrected[,"Genus"])], useNA="ifany")
unname(tax.drp.uncorrected[!bim.drp[!corrected.drp] & is.na(tax.drp.uncorrected[,"Genus"]),])
```

Construct a data.frame of all errors between non-contaminant/non-chimeric reads and the true ASVs.

```{r}
gens <- unique(tax[,"Genus"])
keep <- !is.na(dd$map) # Keep all corrected at default settings
keep[!keep] <- tax.drp.uncorrected[,"Genus"] %in% gens # Also keep uncorrected if in the expected genera
keep[bim.drp] <- FALSE # Don't keep chimeras
ii <- which(keep) # Index version of keep
diffs.keep <- lapply(ii, df.diffs) # ~2 mins
bar <- do.call(rbind, diffs.keep)
bar <- bar[order(bar$Qual, decreasing=TRUE),]
```

Count up the number of reads that were kept that extend to each nucleotide position.
```{r}
nnt.keep <- sapply(seq(max(nchar(getSequences(drp)[keep]))), function(pos) {
  sum(drp$uniques[keep & nchar(getSequences(drp)) >= pos])
})
```

Barplots with quality colored as a function of query sequence position:
```{r}
pbar <- bar
pbar <- pbar[!is.na(pbar$Qual),]
pbar$Quality <- cut(pbar$Qual, c(0, 50, 70, 90, 93))
qual.map <- c("(0-50]"="0-50", "(51,70]"="51-70", "(70,90]"="71-90", "(70,90]"="91+")
pbar$Quality <- qual.map[pbar$Quality]
pbar$Quality[pbar$Type=="D"] <- "NA"
pbar$Quality <- factor(pbar$Quality, levels=c(qual.map, "NA"))
type.map <- c("S"="Substitution", "D"="Deletion", "I"="Insertion")
pbar$Error <- factor(type.map[pbar$Type], levels=c("Substitution", "Insertion", "Deletion"))
color.scale <- c("hotpink", colorRampPalette(c("deeppink2", "dodgerblue2"))(3))
color.scale <- c(colorRampPalette(c("black", "cyan"))(4), "grey")
# Convert counts to rates by using the total lengths of all reads
pbar$Rate <- pbar$Abund/nnt.keep[pbar$QueryPos]
# Force desired facet ymax limits with a custom data.frame, and set desired breaks
dflim <- data.frame(Rate=c(0.001, 0.01, 0.003), 
                    QueryPos=c(200, 200, 200), 
                    Quality=c("NA", "NA", "NA"),
                    Error=c("Substitution", "Insertion", "Deletion"))
my_breaks <- function(x) { 
  if (max(x) < 0.0015) { c("0.0000"=0, "0.0010"=0.001) }  # "0.0005"=0.0005, 
  else if(max(x) < 0.005) { c("0.0000"=0, "0.0010"=0.001, "0.0020"=0.002, "0.0030"=0.003) }
#  else { c("0.0000"=0, "0.0050"=0.005, "0.0100"=0.01) }
  else { c("0.0000"=0, "0.0020"=0.002, "0.0040"=0.004, "0.0060"=0.006, "0.0080"=0.008, "0.0100"=0.01) }
}
p.err.pos <- ggplot(data=pbar, aes(x=QueryPos,y=Rate,color=NULL, fill=Quality)) + geom_col() +
  facet_grid(Error~., scales="free_y") + guides(color=FALSE) + xlab("Nucleotide Position") + ylab("Error Rate") +
  scale_color_manual(values=color.scale) + scale_fill_manual(values=color.scale) +
  geom_blank(data=dflim) + scale_y_continuous(breaks=my_breaks) + theme(axis.text.y=element_text(size=7))
p.err.pos
ggsave(file.path(path.out, "Zymo_errpos.pdf"), p.err.pos, width=178, height=100, units="mm", useDingbats=FALSE)
p.err.pos.200 <- ggplot(data=pbar[pbar$QueryPos %in% seq(151,350),], aes(x=QueryPos,y=Rate, color=NULL, fill=Quality)) + geom_col() +
  facet_grid(Error~., scales="free_y") + guides(color=FALSE) + xlab("Nucleotide Position") + ylab("Error Rate") +
  scale_color_manual(values=color.scale) + scale_fill_manual(values=color.scale) +
  geom_blank(data=dflim) + scale_y_continuous(breaks=my_breaks) + theme(axis.text.y=element_text(size=7))
p.err.pos.200
ggsave(file.path(path.out, "Zymo_errpos_200.pdf"), p.err.pos.200, width=178, height=78, units="mm", useDingbats=FALSE)
ggsave(file.path(path.out, "Zymo_errpos_200.png"), p.err.pos.200, dpi=300)
#  scale_color_gradientn(colors=rainbow(5)) + facet_wrap(~Type)
# In final polishing, used Illustrator toadd the intermediate ticks to the Insertion facet as well
# so the ticks are the same scale (0.001) in each plot
```

That is pretty cool. The two big peaks in substitutionss are strongly associated with lower quality scores, and quality is very informative about insertions! Quality doesn't work for deletions of course because deletions as they aren't in the query sequence, and hence don't really have an associated quality score.

Calculate aggregate error probabilities for total errors and each type individually:
```{r}
sq.drp <- getSequences(drp)
totbases.keep <- sum( (nchar(sq.drp)*drp$uniques)[keep] )
tapply(bar$Abund, bar$Type, sum); sum(bar$Abund)
tapply(bar$Abund, bar$Type, sum)/totbases.keep; sum(bar$Abund)/totbases.keep
```

Plot error rates versus quality scores for subsitutions and insertions.
```{r}
srq <- readFastq(filt)
qmat <- as(quality(srq), "matrix") # Raw quality scores
totqs.keep <- tabulate(as.vector(qmat[keep,])) # entries 1-93 for the Q scores
imat.subs <- cbind(drpi = unlist(sapply(bar$Derep[bar$Type=="S"], function(i) which(drp$map==i))),
                   pos = rep(bar$QueryPos[bar$Type=="S"], times=bar$Abund[bar$Type=="S"]))
subqs.keep <- tabulate(as.vector(qmat[imat.subs]))
imat.ins <- cbind(drpi = unlist(sapply(bar$Derep[bar$Type=="I"], function(i) which(drp$map==i))),
                  pos = rep(bar$QueryPos[bar$Type=="I"], times=bar$Abund[bar$Type=="I"]))
insqs.keep <- tabulate(as.vector(qmat[imat.ins]))
```

```{r}
df.qplot <- data.frame(Quality=rep(seq(93),2), 
                       Count=c(subqs.keep, insqs.keep), 
                       Rate=c(subqs.keep/totqs.keep, insqs.keep/totqs.keep),
                       Type=rep(c("Substitution", "Insertion"), each=93))
ggplot(data=df.qplot, aes(x=Quality, y=Rate, color=Type)) + 
  geom_hline(aes(yintercept=sum(bar$Abund[bar$Type=="D"])/totbases.keep), color="grey", linetype="dashed") +
  geom_point() + 
  scale_y_log10(breaks=c("0.0001"=0.0001, "0.001"=0.001, "0.01"=0.01, "0.1"=0.1), name="Error Rate") + 
  scale_color_manual(values=c(Insertion="maroon", Substitution="orange"), name="Error Type") +
  scale_x_continuous(breaks=c(0, 20, 40, 60, 80, 93), limits=c(0, 93), name="Quality Score") +
  theme_bw()
```

The saturation of substitution errors at high Q are probably explained by PCR. The complementation symmetry especially at high Q values in the `plotErrors` further suggests PCR origin for most of these errors.

```{r}
plotErrors(err)
```

Some summary statistics and plots on the distribution of quality scores:
```{r}
totqs.keep[[93]]/sum(totqs.keep) # Fraction with max q=93 quality score
ggplot(data=data.frame(Quality=seq(93), Count=totqs.keep), aes(x=Quality, y=Count)) +
  geom_step() + theme_bw() + scale_y_log10()
ggplot(data=data.frame(Quality=seq(93), Cumulative=cumsum(totqs.keep)/sum(totqs.keep)), aes(x=Quality, y=Cumulative)) +
  geom_step() + theme_bw()
```

Also calculate the number of CCS reads that were error-free at various stages.
```{r}
n.error.free <- sum(drp$uniques[dd$sequence])
cbind(ccs=prim[,1], primers=prim[,2], filtered=track[[2]], denoised=sum(dd$denoised), error.free=n.error.free)
n.error.free/prim[[1]] # Fraction error-free out of all CCS reads
n.error.free/track[[1]] # Fraction error-free out of CCS reads with detected primers, but before filtering
```

