---
title: "LRASms_ATCC"
author: "BJC"
date: "8/31/2018"
output: html_document
---

## Data

>The ATCC samples were sequenced using the same chemistry as the first run of the fecal samples.  The 16S CCS sequences at >99.9% accuracy are labeled:
>
>The two cells only differed in loading concentration on the Sequel.

https://www.atcc.org/Products/All/MSA-1003.aspx#generalinformation

Components	
0.18% Acinetobacter baumannii (ATCC 17978)
0.02% Actinomyces odontolyticus (ATCC 17982)
1.80% Bacillus cereus (ATCC 10987)
0.02% Bacteroides vulgatus (ATCC 8482)
0.02% Bifidobacterium adolescentis (ATCC 15703)
1.80% Clostridium beijerinckii (ATCC 35702)
0.02% Deinococcus radiodurans (ATCC BAA-816)
0.02% Enterococcus faecalis (ATCC 47077)
18.0% Escherichia coli (ATCC 700926)
0.18% Helicobacter pylori (ATCC 700392)
0.18% Lactobacillus gasseri (ATCC 33323)
0.18% Neisseria meningitidis (ATCC BAA-335)
18.0% Porphyromonas gingivalis (ATCC 33277)
0.18% Propionibacterium acnes (ATCC 11828)
1.80% Pseudomonas aeruginosa (ATCC 9027)
18.0% Rhodobacter sphaeroides (ATCC 17029)
1.80% Staphylococcus aureus (ATCC BAA-1556)
18.0% Staphylococcus epidermidis (ATCC 12228)
1.80% Streptococcus agalactiae (ATCC BAA-611)
18.0% Streptococcus mutans (ATCC 700610)

This is essentially the HMP mock community again, albeit from ATCC and with some small changes relative to the BEI formulation. As these samples were run on full Sequel cells with the S/P2-C2/5.0 sequencing chemistry, they are much deeper than the BEI HMP mock run.

## Setup

Load libraries, setup paths, prepare environment:
```{r init, warning=FALSE, message=FALSE}
library(dada2);packageVersion("dada2")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
library(reshape2); packageVersion("reshape2")
path <- "~/Desktop/LRAS/Data/ATCC/" # CHANGE ME to location of the fastq file
path.out <- "Figures/" 
path.rds <- "RDS/" 
fn1 <- file.path(path, "ATCC_16S_CCS99.9_cell1.fastq.gz")
fn2 <- file.path(path, "ATCC_16S_CCS99.9_cell2.fastq.gz")
F27 <- "AGRGTTYGATYMTGGCTCAG"
R1492 <- "RGYTACCTTGTTACGACTT"
rc <- dada2:::rc
theme_set(theme_bw())
```

Remove primers and orient reads:
```{r primers}
nop1 <- file.path(path, "noprimers", basename(fn1))
nop2 <- file.path(path, "noprimers", basename(fn2))
dada2:::removePrimers(fn1, nop1, primer.fwd=F27, primer.rev=dada2:::rc(R1492), orient=TRUE, verbose=TRUE)
dada2:::removePrimers(fn2, nop2, primer.fwd=F27, primer.rev=dada2:::rc(R1492), orient=TRUE, verbose=TRUE)
```

Takes like 20 minutes, suggests we should revisit the speed on this. That said, a lot of reads in these (~300k) and a high fraction passing the primer check (~85%).

Inspect length distribution.
```{r}
hist(nchar(getSequences(nop1)), 100)
hist(nchar(getSequences(nop2)), 100)
```

Peaked in the right place, some very long-length outliers.

Filter:
```{r filter}
filt1 <- file.path(path, "noprimers", "filtered", basename(fn1))
filt2 <- file.path(path, "noprimers", "filtered", basename(fn2))
track <- filterAndTrim(c(nop1, nop2), c(filt1, filt2), minQ=3, minLen=1000, maxLen=1600, maxN=0, rm.phix=FALSE, maxEE=2, verbose=TRUE)
```

High rate (~96-97%) of reads passing the filter.

## Run DADA2

Dereplicate:
```{r derep}
drp1 <- derepFastq(filt1, verbose=TRUE)
drp2 <- derepFastq(filt2, verbose=TRUE)
```

Learn errors:

```{r learn-err}
err1 <- learnErrors(drp1, BAND_SIZE=32, multithread=TRUE, errorEstimationFunction=dada2:::PacBioErrfun)
err2 <- learnErrors(drp2, BAND_SIZE=32, multithread=TRUE, errorEstimationFunction=dada2:::PacBioErrfun)
```

Takes over half-an-hour.

Inspect errors:
```{r plot-err}
plotErrors(err1)
plotErrors(err2)
```

Looks good. And pretty identical.

Denoise:
```{r dada}
dd1 <- dada(drp1, err=err1, BAND_SIZE=32, multithread=TRUE) # seconds
dd2 <- dada(drp2, err=err2, BAND_SIZE=32, multithread=TRUE) # seconds
dd1
dd2
```

Inspect clustering data.frames
```{r inspect}
dd1$clustering[,-1]
dd2$clustering[,-1]
```

## Taxonomy and Chimeras

Assign taxonomy:
```{r taxonomy}
tax1 <- assignTaxonomy(dd1, "~/tax/silva_nr_v128_train_set.fa.gz", multithread=TRUE)
tax1[,"Genus"] <- gsub("Escherichia/Shigella", "Escherichia", tax1[,"Genus"]) # Reformat to be compatible with other data sources
tax2 <- assignTaxonomy(dd2, "~/tax/silva_nr_v128_train_set.fa.gz", multithread=TRUE)
tax2[,"Genus"] <- gsub("Escherichia/Shigella", "Escherichia", tax2[,"Genus"]) # Reformat to be compatible with other data sources
```

Check chimeras:
```{r chimeras}
bim1 <- isBimeraDenovo(dd1, minFoldParentOverAbundance=3.5, multithread=TRUE) 
# Higher MFPOA to avoid flagging intra-genomic variants
table(bim1)
bim2 <- isBimeraDenovo(dd2, minFoldParentOverAbundance=3.5, multithread=TRUE) 
# Higher MFPOA to avoid flagging intra-genomic variants
table(bim2)
```

Lots of chimeras! Very different from the other mock communities! Check with PacBio about the PCR protocol used on these samples.

Save processed objects for future analysis.
```{r saveRDS}
saveRDS(err1, file.path(path.rds, "ATCC_err1.rds"))
saveRDS(dd1, file.path(path.rds, "ATCC_dd1.rds"))
saveRDS(bim1, file.path(path.rds, "ATCC_bim1.rds"))
saveRDS(tax1, file.path(path.rds, "ATCC_tax1_Silva128.rds"))
saveRDS(err2, file.path(path.rds, "ATCC_err2.rds"))
saveRDS(dd2, file.path(path.rds, "ATCC_dd2.rds"))
saveRDS(bim2, file.path(path.rds, "ATCC_bim2.rds"))
saveRDS(tax2, file.path(path.rds, "ATCC_tax2_Silva128.rds"))
```

Reload processed data objects (can run code below from here in absence of input sequence data):
```{r readRDS}
dd1 <- readRDS(file.path(path.rds, "ATCC_dd1.rds"))
bim1 <- readRDS(file.path(path.rds, "ATCC_bim1.rds"))
tax1 <- readRDS(file.path(path.rds, "ATCC_tax1_Silva128.rds"))
dd2 <- readRDS(file.path(path.rds, "ATCC_dd2.rds"))
bim2 <- readRDS(file.path(path.rds, "ATCC_bim2.rds"))
tax2 <- readRDS(file.path(path.rds, "ATCC_tax2_Silva128.rds"))
```

Build a joint sequence table:
```{r}
sta <- makeSequenceTable(list(Rep.1=dd1, Rep.2=dd2))
st <- removeBimeraDenovo(sta, minFoldParentOverAbundance=3.5, multithread=TRUE, verbose=TRUE)
```

Take a look at the correspondence between abundances in these samples:
```{r}
table(st["Rep.1",]>0, st["Rep.2",]>0)
tapply(colSums(st), list(st["Rep.1",]>0, st["Rep.2",]>0), sum)
plot(st["Rep.1",], st["Rep.2",])
plot(st["Rep.1",], st["Rep.2",], log="xy")
```

Correspondence is quite good. 8 variants in each that aren't in the other and with very similar numbers of reads (but relatively few). May want to dig into that a bit more.

Assign taxonomy:
```{r}
tax <- assignTaxonomy(st, "~/tax/silva_nr_v128_train_set.fa.gz", multithread=TRUE)
tax[,"Genus"] <- gsub("Escherichia/Shigella", "Escherichia", tax[,"Genus"]) # Reformat to be compatible with other data sources
unname(tax)
```

Definitely are extra Escherichia floating around...

```{r}
cbind(dd1$clustering[match(colnames(st), dd1$sequence),-1], seq(ncol(st)))
```





